### PR TITLE
Knative integration

### PR DESCRIPTION
**Pull request by shyamjesal**
- add proxy to mimic knative QP behavior
- moved unit and integration tests to `tests`
- removed DST timeout tests. 
  - QP is reponsible for swawning the dstFn.
- start grpc server without interceptor if tracing is diabled
- removed fnInvocation proto

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -6,6 +6,8 @@
   "DstServerPort": ":50007",
   "DstServerHostname": "localhost",
   "SQPServerPort": ":50005",
+  "ProxyHostname": "localhost",
```
Comment (ustiugov) [inline config.json:9]: Need both dQP and this?

File: integration-tests/integration_test.go [hunk 2]
```diff
@@ -61,6 +68,55 @@ var handler = func(data []byte) {
...
+
+	httpProxy := func(target string) *httputil.ReverseProxy {
+		return &httputil.ReverseProxy{
+			Director: func(req *http.Request) {
+				req.URL.Scheme = "http"
+				req.URL.Host = target
+
+				// Copied from httputil.NewSingleHostReverseProxy.
+				if _, ok := req.Header["User-Agent"]; !ok {
+					// explicitly disable User-Agent so it's not set to default value
```
Comment (ustiugov) [inline integration-tests/integration_test.go]: Why?

File: integration-tests/integration_test.go [hunk 3]
```diff
@@ -61,6 +68,55 @@ var handler = func(data []byte) {
...
+					req.Header.Set("User-Agent", "")
+				}
+			},
+		}
+	}(config.DstServerHostname + config.DstServerPort)
+	httpProxy.Transport = pkgnet.NewProxyAutoTransport(10 /* max-idle */, 5 /* max-idle-per-host */)
```
Comment (ustiugov) [inline integration-tests/integration_test.go]: What is that?

File: integration-tests/integration_test.go [hunk 4]
```diff
@@ -61,6 +68,55 @@ var handler = func(data []byte) {
...
+
+	var composedHandler http.Handler = httpProxy
+
+	composedHandler = func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			defer h.ServeHTTP(w, r)
+
+			isXDT := r.Header.Get("is_xdt")
+			if isXDT == "true" {
+				log.Infof("pulling from sQP using key %s addr %s", r.Header.Get("key"), r.Header.Get("sqp_addr"))
```
Comment (ustiugov) [inline integration-tests/integration_test.go]: Need to think how to support multiple objects, please think of the API

File: integration-tests/integration_test.go [hunk 5]
```diff
@@ -79,23 +135,23 @@ func TestSdk_InvokeWithXDT(t *testing.T) {
...
-	go dQP.StartServer(config)
-	go StartDstServer(config, handler)
+	go knativeQP(config)
+	go sdk.StartDstServer(config, handler)
 
 	time.Sleep(time.Second * 1)
 
 	start := time.Now()
 	log.Infof("starting integ test")
-	url := config.DQPServerHostname + config.DQPServerPort
```
Comment (ustiugov) [inline integration-tests/integration_test.go:97]: Are you going to delete dQP completely?

File: integration-tests/integration_test.go [hunk 6]
```diff
@@ -134,42 +190,11 @@ func TestErr_DQPTimeout(t *testing.T) {
 	log.Infof("completed XDT in %s", elapsed)
 }
 
-func TestErr_DSTTimeout(t *testing.T) {
```
Comment (ustiugov) [inline integration-tests/integration_test.go:137]: Why delete this test?

File: integration-tests/unit_test.go [hunk 7]
```diff
@@ -20,7 +20,7 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
```
Comment (ustiugov) [inline integration-tests/unit_test.go]: Unit test named file is a contradiction when located in integration tests folder

File: sdk/source.go [hunk 8]
```diff
@@ -54,17 +56,23 @@ func splitPayload(xdtPayload *utils.Payload) (string, []byte) {
...
 	if err != nil {
 		return err
 	}
 
+	httpMetadata := map[string]string{
+		"is_xdt":   "true",
+		"key":      key,
+		"sqp_addr": sQPAddr,
+	}
+	ctx := metadata.NewOutgoingContext(context.Background(), metadata.New(httpMetadata))
```
Comment (ustiugov) [inline sdk/source.go:70]: You create a new context here, maybe this is what breaks tracing?

File: user-functions/dQP/dQP.go [hunk 9]
```diff
@@ -24,13 +24,20 @@ package main
```
Comment (ustiugov) [inline user-functions/dQP/dQP.go:24]: Decide to keep dQP or remove it, better to remove it if not used anymore

File: user-functions/gx/go.mod [hunk 10]
```diff
@@ -16,6 +16,7 @@ replace (
 )
```
Comment (ustiugov) [inline user-functions/gx/go.mod:17]: Clean up replaces above

**Review by ustiugov — CHANGES_REQUESTED** [review]
Cool! 

Remove dQP if unused, more comments inline

File: config.json (hunk 1 above)
Comment (shyamjesal) [reply config.json:9]: We had 3 servers. sQP, dQP and dstFn. The dQP server used to perform 2 tasks, routing the request and pre-fetching the payload. Now the routing is done by the proxy server. Hence we require both.

File: integration-tests/integration_test.go (hunk 2 above)
Comment (shyamjesal) [reply integration-tests/integration_test.go]: These lines are copied from Knative code-base as-is.

File: integration-tests/integration_test.go (hunk 3 above)
Comment (shyamjesal) [reply integration-tests/integration_test.go]: This transport allows forwarding of binary type gPPC data using http proxy.
Without this, the proxy tries inferring the data as simple text and results in error.

File: integration-tests/integration_test.go (hunk 6 above)
Comment (shyamjesal) [reply integration-tests/integration_test.go:137]: Before implementing the proxy, we could spawn dQP (the routing part) without having dstFn. 
But, after having the reverse proxy, this is no longer possible. The proxy quits with an error saying destination not found.

File: integration-tests/integration_test.go (hunk 3 above)
Comment (ustiugov) [reply integration-tests/integration_test.go]: Please clarify the constants' role and what the values are and why
Add a comment

File: sdk/source.go (hunk 8 above)
Comment (ustiugov) [reply sdk/source.go:70]: actually, i don't think so...

File: tests/go.mod [hunk 11]
```diff
@@ -0,0 +1,30 @@
+module github.com/ease-lab/xdt/integration_tests
+
+go 1.16
+
+replace (
+	XDTgRPC_stream/plotter => ../plotter
+	github.com/ease-lab/xdt/dQP => ../dQP
+	github.com/ease-lab/xdt/proto/crossXDT => ../proto/crossXDT
+	github.com/ease-lab/xdt/proto/downXDT => ../proto/downXDT
+	github.com/ease-lab/xdt/proto/fnInvocation => ../proto/fnInvocation
```
Comment (ustiugov) [inline tests/go.mod]: remove

File: tests/integration_test.go [hunk 12]
```diff
@@ -61,6 +68,60 @@ var handler = func(data []byte) {
...
+
+				// Copied from httputil.NewSingleHostReverseProxy.
+				if _, ok := req.Header["User-Agent"]; !ok {
+					// explicitly disable User-Agent so it's not set to default value
+					req.Header.Set("User-Agent", "")
+				}
+			},
+		}
+	}(config.DstServerHostname + config.DstServerPort)
+	httpProxy.Transport = pkgnet.NewProxyAutoTransport(10 /* max-idle */, 5 /* max-idle-per-host */)
```
Comment (ustiugov) [inline tests/integration_test.go]: figure this out and add a comment please. what are these constants stand for?
hint from the http/net module that I've got from google in 2 min:
```
    // MaxIdleConns controls the maximum number of idle (keep-alive)
    // connections across all hosts. Zero means no limit.
    MaxIdleConns int // Go 1.7

    // MaxIdleConnsPerHost, if non-zero, controls the maximum idle
    // (keep-alive) connections to keep per-host. If zero,
    // DefaultMaxIdleConnsPerHost is used.
    MaxIdleConnsPerHost int
```
    
basically we may see re-connections for functions that connect to too many sources. probably leave as is but add a comment

File: tests/integration_test.go [hunk 13]
```diff
@@ -61,6 +68,60 @@ var handler = func(data []byte) {
...
+
+	composedHandler = func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			defer h.ServeHTTP(w, r)
+
+			isXDT := r.Header.Get("is_xdt")
+			if isXDT == "true" {
+				log.Infof("pulling from sQP using key %s addr %s", r.Header.Get("key"), r.Header.Get("sqp_addr"))
+				go func() {
+					err := dQP.PullDataFromSrcQP(r.Context(), r.Header.Get("key"), r.Header.Get("sqp_addr"), config.ChunkSizeInBytes)
```
Comment (ustiugov) [inline tests/integration_test.go:109]:
```suggestion
// FIXME: support many payloads per invocation
					err := dQP.PullDataFromSrcQP(r.Context(), r.Header.Get("key"), r.Header.Get("sqp_addr"), config.ChunkSizeInBytes)
```

File: tests/integration_test.go [hunk 14]
```diff
@@ -61,6 +68,60 @@ var handler = func(data []byte) {
...
+				}()
+			}
+
+		})
+	}(composedHandler)
+
+	composedHandler = queue.ForwardedShimHandler(composedHandler)
+
+	h2s := &http2.Server{}
+	// start server
```
Comment (ustiugov) [inline tests/integration_test.go]: remove

**Review by ustiugov — COMMENTED** [review]
please describe the changes in the PR's description

then you can merge it after the cosmetic changes I requested

**Pull request merged**